src/dotnet/SlnFileFactory.cs

```diff
@@ -0,0 +1,86 @@
+﻿// Copyright (c) .NET Foundation and contributors. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.IO;
+using System.Linq;
+using Microsoft.DotNet.Cli.Sln.Internal;
+using Microsoft.DotNet.Cli.Utils;
+
+namespace Microsoft.DotNet.Tools.Common
+{
+    public static class SlnFileFactory
+    {
+        public static SlnFile CreateFromFileOrDirectory(string fileOrDirectory)
+        {
+            if (File.Exists(fileOrDirectory))
+            {
+                return FromFile(fileOrDirectory);
+            }
+            else
+            {
+                return FromDirectory(fileOrDirectory);
+            }
+        }
+
+        private static SlnFile FromFile(string solutionPath)
+        {
+            SlnFile slnFile = null;
+            try
+            {
+                slnFile = SlnFile.Read(solutionPath);
+            }
+            catch
+            {
```

Comment:
I wanted to provide a general error in case of any failure. We also have unit tests for the SlnFile reader (and I have an issue open to add more comprehensive tests). So basically whenever we cannot open the file for now we just say invalid solution until we have the ability to detect specific cases.